### PR TITLE
Add error checking for when tl doesn't catch lua errors

### PR DESCRIPTION
**Pull request by euclidianAce**
In addition, use `tl.load` to load modules since the code for both was very similar.

### CODE REVIEW

**Comment by euclidianAce** [comment]
Fixes #216

File: tl.tl [hunk 1]
```diff
@@ -6509,17 +6517,16 @@ local function tl_package_loader(module_name: string): any
    if found_filename then
       local input = fd:read("*a")
       fd:close()
-      local tokens = tl.lex(input)
-      local errs = {}
-      local i, program = tl.parse_program(tokens, errs, found_filename)
-      local code = tl.pretty_print_ast(program, true)
-      local chunk = load(code, found_filename)
+      local chunk, err = tl.load(input, module_name, "t")
```
Comment (hishamhm) [inline tl.tl:6520]: I'm not sure by just looking at the diff, but: shouldn't it be found_filename here instead of module_name? So that any parse errors contain the correct filename

Comment (euclidianAce) [reply tl.tl:6520]: Ah, yes it should.

File: tl.tl [hunk 2]
```diff
@@ -6509,17 +6517,16 @@ local function tl_package_loader(module_name: string): any
...
+      local chunk, err = tl.load(input, found_filename, "t")
       if chunk then
          return function(): any
             local ret = chunk()
             package.loaded[module_name] = ret
             return ret
          end
+      else
+         io.stderr:write("!!Error uncaught by tl!!\n", err, "\n\n\tThis is probably a bug caused by tl not catching invalid Lua.\n\tPlease consider reporting this at https://github.com/teal-language/tl\n\n")
+	 os.exit(1)
```
Comment (hishamhm) [inline tl.tl:6529]: On a second reading, I'm not sure this is right. When tl.load _does_ catch a parse error, it will return `nil, err` normally, so this not an "error uncaught by tl".

In the case of `type_check_and_load`, the situation is different: that "error uncaught by tl" message refers to code generated by tl from an AST produced by `tl.process` which finished without catching errors. (That is, an error-free AST shouldn't be generating code that the Lua VM fails to `load`).

Looking at the original code, perhaps what we really need is to handle the errors of `tl.parse_program` (Teal caught an error) and `load` (Lua caught an error that Teal has let slip by) separately.

Comment (euclidianAce) [reply tl.tl:6529]: Ah, I see what you mean. Something like this should (hopefully) be right.
```lua
local function tl_package_loader(module_name: string): any
   local found_filename, fd, tried = tl.search_module(module_name, false)
   if found_filename then
      local input = fd:read("*a")
      fd:close()
      local errs: {ParseError} = {}
      local _, program = tl.parse_program(tl.lex(input), errs, module_name)
      if #errs > 0 then
         error(module_name .. ":" .. errs[1].y .. ":" .. errs[1].x .. ": " .. errs[1].msg)
      end
      local code = tl.pretty_print_ast(program, true)
      local chunk, err = load(code, module_name, "t")
      if chunk then
         return function(): any
            local ret = chunk()
            package.loaded[module_name] = ret
            return ret
         end
      else
         io.stderr:write("!!Error uncaught by tl!!\n", err, "\n\n\tThis is probably a bug caused by tl not catching invalid Lua.\n\tPlease consider reporting this at https://github.com/teal-language/tl\n\n")
         os.exit(1) -- maybe error here instead as I'm not sure how os.exit interacts with tests
      end
   end
   return table.concat(tried, "\n\t")
end
```
But I'm not too certain of a good way to test this without relying on Teal not behaving correctly.
Perhaps `tl.parse_program` could be overridden to never error and pass bad code to `load`?
Something like:
```lua
local tl = require("tl")
describe("Error checking", function()
   local old_parse_program
   setup(function()
      old_parse_program = tl.parse_program
      tl.parse_program = function(tokens, _, chunkname)
         return old_parse_program(tokens, {}, chunkname)
      end
   end)
   teardown(function()
      tl.parse_program = old_parse_program
   end)

   it("should report when Teal doesn't catch a Lua error", function()
      -- [[ ... ]]
   end)
end)

```

Comment (hishamhm) [reply tl.tl:6529]: Agree that `error()` is better than `os.exit`. Being a "library" module, the `tl.tl` file must never `os.exit`.

About the testing, this sounds like a good strategy. But I wouldn't sweat a lot if this is too hard to test. It's very much a corner case. 

Also, I would change the error message to `Internal Compiler Error: Teal generator produced invalid Lua. Please report a bug at https://github.com/teal-language/tl`

**Comment by euclidianAce** [comment]
Alright, error messages changed and a small test added. Seems to work on my end. Let me know if it needs anything else!
